Pileup-related changes (LIBS/LocusWalker/ReadPileup)

magicDGS opened the pull request:
This PR tries to solve several issues:
* Refactoring constructors for LIBS (#1879)
* Adding maximum depth per sample argument to `LocusWalker` to avoid memory overload
* Fixing `Pileup`/`CheckPileup` tools for command line read filters
* Method for fix overlaps in `ReadPileup` in the same way as samtools (#2034)

magicDGS commented:
@droazen , I guess that this will be for you because you were in the previous PR ( #2041). I think that this is cleaner and does not interfere in the basic behavior for GATK tools; but it will be very useful for my work.

Thanks a lot!

droazen commented:
Re-assigning to @fleharty

magicDGS commented:
Hi @fleharty! Just a friendly ping here, because I need this in my work.

fleharty commented:
@magicDGS reviewing

fleharty commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -64,6 +66,10 @@
     @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "Output file (if not provided, defaults to STDOUT).", optional = true)
     public File outFile = null;
 
+    // this is to check for previous versions of samtools without having overlap detection
```
Capitalize This

magicDGS replied:
Done

fleharty commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -72,6 +78,16 @@
     @Argument(fullName="continue_after_error", doc="Continue after encountering an error", optional=true)
     public boolean continueAfterAnError = false;
 
+    // this are the default read filters from samtools
```
this -> These

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByState.java:
```diff
@@ -33,7 +33,7 @@
  * -- Only aligned reads with at least one on-genomic cigar operator are passed on in the pileups.  That is,
  * unmapped reads or reads that are all insertions (10I) or soft clipped (10S) are not passed on.
- * -- LIBS can perform per-sample downsampling of a variety of kinds.
+ * -- LIBS can perform per-sample downsampling of a variety of kinds. TODO: it is true?
```
Should this still be here?  I think this TODO should be removed.

magicDGS replied:
I added this TODO here while doing this changes because I found that the per-sample downsampling is using `LevelingDownsampler`, so I don't know if it is true that it performs dowsampling "of a variety of kinds", even if the `DownsamplingMethod` passed to the constructor is different.

droazen replied:
Just adjust the comment to read "LIBS can perform per-sample downsampling to a coverage target if requested", and remove the TODO.

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByState.java:
```diff
@@ -95,115 +95,118 @@
-     * @param downsamplingInfo meta-information about how to downsampling the reads
+     * @param samIterator                    the iterator of reads to process into pileups.  Reads must be ordered
+     *                                       according to standard coordinate-sorted BAM conventions
+     * @param downsamplingInfo               meta-information about how to downsampling the reads
```
downsampling -> downsample

magicDGS replied:
Done

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByState.java:
```diff
@@ -95,115 +95,118 @@
+     * @param samIterator                    the iterator of reads to process into pileups.  Reads must be ordered
+     *                                       according to standard coordinate-sorted BAM conventions
+     * @param downsamplingInfo               meta-information about how to downsampling the reads
+     * @param keepUniqueReadListInLIBS       if true, we will keep the unique reads from off the samIterator and make them
```
from "off" ?

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -6,6 +6,8 @@
 import org.broadinstitute.hellbender.utils.BaseUtils;
 import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.fragments.FragmentCollection;
+import org.broadinstitute.hellbender.utils.fragments.FragmentUtils;
```
FragmentUtils is not needed, please remove

fleharty commented on src/test/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByStateBaseTest.java:
```diff
@@ -64,11 +64,9 @@ protected LocusIteratorByState makeLIBS(final List<GATKRead> reads,
-                keepUniqueReadList,
-                sampleListForSAMWithoutReadGroups(),
-                header);
+                keepUniqueReadList, sampleListForSAMWithoutReadGroups(), header, true,
```
Use of new lines is inconsistent, please fix.

magicDGS replied:
Sorry, it was because of the IDE refactoring. Fixed.

fleharty commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java:
```diff
@@ -490,4 +491,106 @@ public void testSimplePileupWithOffset(){
+        read2.setBaseQualities(new byte[] {lowQual, highQual, lowQual, highQual,
+                SAMUtils.MAX_PHRED_SCORE});
+        return new Object[][] {
+                {PileupElement.createPileupForReadAndOffset(read1, 0),
```
For each of these tests, could you provide a 1 sentence description of what that test is establishing?

magicDGS replied:
Done

fleharty commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java:
```diff
@@ -490,4 +491,106 @@ public void testSimplePileupWithOffset(){
+            Assert.assertEquals(element3.getQual(), Byte.MAX_VALUE);
+            Assert.assertEquals(element1.getQual(), 0);
+        }
+        // finally, check what happen if both are the maximum value
```
happen -> happens
finally -> Finally

magicDGS replied:
Done

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +239,69 @@ public Locatable getLocation() {
+                // if not, we lost confidence in the one with higher quality
+                if (firstElement.getQual() >= secondElement.getQual()) {
+                    firstQuals[firstElement.getOffset()] =
+                            (byte) (0.8 * firstQuals[firstElement.getOffset()]);
```
I'm pretty uncomfortable with this 0.8, where does it come from?
I realize that it is in https://github.com/samtools/htslib/blob/master/sam.c
but I also don't see justification there either.

magicDGS replied:
I also don't like it, but because I want to check how consistent it is with the samtools result, I definitely require that this is in sync with the sam.c code.

magicDGS replied:
A suggestion for make this better will be to add a `double` parameter to downgrade the confidence of a mismatch. Like that I can use it with 0.8 to mimic samtools behaviour (and change it if the samtools method is updated), and other API users could use more reasonable values (e.g. PCR error rate correction and/or just set to 0 to do not take it into account).

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +239,69 @@ public Locatable getLocation() {
+                    secondQuals[secondElement.getOffset()] = 0;
+                } else {
+                    secondQuals[secondElement.getOffset()] =
+                            (byte) (0.8 * secondQuals[secondElement.getOffset()]);
```
Here too.

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +239,69 @@ public Locatable getLocation() {
+     *
+     * 1. If the bases are the same, the quality of the first element is the sum of both qualities
+     *    and the quality of the second is reduced to 0.
+     * 2. If the bases are different, the base with the highest quality is reduced with a factor of
```
Other than samtools, is there a justification for 0.8?

magicDGS replied:
There is no justification, just the samtools behaviour that I'm trying to mimic here.

My work is related in avoid the step of store the mpileup from samtools, that's why I need this as it is. Actually there are other method in `FragmentUtils` to correct for overlapping read pairs, and probably this will be more suitable for some API users. I can add in the javadoc a link to that method to note that there are other strategies to perform this.

fleharty commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +239,69 @@ public Locatable getLocation() {
+                                .getOffset()]);
+                // cap to maximum byte value
+                if (firstQuals[firstElement.getOffset()] < 0) {
+                    firstQuals[firstElement.getOffset()] = Byte.MAX_VALUE;
```
is Byte.MAX_VALUE really what we want here?  I don't know about GATK4, but in GATK3 there is a util class that has a Maximum sensible base quality.

magicDGS replied:
I was thinking to use the maximum quality in the util class (I think that it is 93), but I decided to use `Byte.MAX_VALUE` to allow the API user to distinguish which values were capped. I don't mind if the value is 93 or `Byte.MAX_VALUE` because anyway we can't mimic the cap value from samtools (200) because byte in Java is signed. Should I change it?

magicDGS replied:
I changed it for `QualityUtils.MAX_SAM_QUAL_SCORE`.

fleharty commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java:
```diff
@@ -490,4 +491,106 @@ public void testSimplePileupWithOffset(){
+        read2.setBaseQualities(new byte[] {40, 40});
+        final GATKRead copy = read.deepCopy();
+        final GATKRead copy2 = read2.deepCopy();
+        final PileupElement deleted = new PileupElement(copy, 1, copy.getCigarElement(1), 1, 0);
```
I'm confused.  copy, copy2, read, and read2 appear to be independent of what is happening in deleted and normal.
Could you clarify what's happening here in a comment?

magicDGS replied:
I will add a comment, but just for clarify it here too: because the method `ReadPileup.fixPairOverlappingQualities()` modify the read inside the `PileupElement`, we need a way to check that anything within the read was changed. Thus, we require a copy to maintain a backup of the original read and check if it was change or not. Is this clear like that, because I guess that is still a bit confusing?

magicDGS commented:
Thanks for the review, @fleharty! I addressed the comments (Travis not passing because of the CLOUD tests).

Regarding the downgrade of quality, I think that it should be 0.8 to keep it as in samtools because it is already noted in the javadoc what we are doing. If someone wants to use a different approach should use other method (like `FragmentUtils.adjustQualsOfOverlappingPairedFragments`) and/or their own implementation. Another way to keep it consistent with samtools but give to the client some freedom, a `double` could be added (checking will be required for the range).

Back to you @fleharty. And thanks again for have a look to this!

fleharty commented:
@magicDGS I'm at a conference right now, I'm going to respond ASAP, hopefully today.

magicDGS commented:
Thanks @fleharty!

magicDGS commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java:
```diff
@@ -575,17 +580,26 @@ public void testFixPairOverlappingQualitiesCap() {
         final PileupElement normal = new PileupElement(copy2, 0, copy.getCigarElement(0), 0, 0);
         ReadPileup.fixPairOverlappingQualities(deleted, normal);
+        // Copy reads are a reference to the read into the PileupElement, and thus should not be modified.
+        // Thus, we use them to test if it is not modified. This could be check uncommenting the next two lines:
```
Done

magicDGS commented:
I addressed the latest comments (using `QualityUtils` static value for capping) and added one more test case. Thanks for the review and back to you @fleharty!

magicDGS commented:
Friendly ping @fleharty! I will really appreaciatte if this can go into before #2185 to avoid the maintenance of the deprecated `PerReadAlleleLikelihoodMap` just for me.

Thanks a lot!

fleharty commented:
@magicDGS Thanks for the changes, looks great 👍

magicDGS commented:
Thanks for reviewing, @fleharty. Could this being merge, @droazen?

davidbenjamin commented:
Restarting the build, which was failing. . .

magicDGS commented:
The build is failing for CLOUD=mandatory, because it always does for people external to Broad. Is that an issue, @davidbenjamin?

davidbenjamin commented:
@magicDGS Ah, got it, @droazen's opinion is the only one that matters.

droazen commented:
Will have a look at this this afternoon

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -30,6 +32,9 @@
  */
 public abstract class LocusWalker extends GATKTool {
 
+    @Argument(fullName = "maxDepthPerSample", shortName = "maxDepthPerSample", doc = "Maximum number of reads to retain per sample. Reads above this threshold will be downsampled. Set to 0 to disable.", optional = true)
```
"per sample" -> "per sample per locus"

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -112,15 +125,9 @@ public CountingReadFilter makeReadFilter(){
-        return DownsamplingMethod.NONE;
+    /** Returns the downsampling info using {@link #maxDepthPerSample} as target coverage. */
+    protected final LIBSDownsamplingInfo getDownsamplingInfo() {
+        return new LIBSDownsamplingInfo(maxDepthPerSample != 0, maxDepthPerSample);
```
set the second argument to the `LIBSDownsamplingInfo` to -1 in the case where we're not downsampling, as per the method javadoc

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/CheckPileup.java:
```diff
@@ -64,6 +66,10 @@
     public File outFile = null;
 
+    // This is to check for previous versions of samtools without having overlap detection
+    @Argument(fullName = "ignore_overlaps", shortName = "x", doc = "Disable read-pair overlap detection.", optional = true)
```
`x` is not a good `shortName` for this argument -- recommend setting `shortname` to `ignore_overlaps` as well.

magicDGS replied:
I used the sames as in samtools mpileup, but it is true that it is not a good short name. I changed it.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +240,71 @@ public Locatable getLocation() {
+    }
+
+    /**
+     * Fixes the quality of two elements that come from an overlaping pair in the same way as
```
overlaping -> overlapping

magicDGS replied:
fixed

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -237,6 +240,71 @@ public Locatable getLocation() {
+                            (byte) (0.8 * firstQuals[firstElement.getOffset()]);
+                    secondQuals[secondElement.getOffset()] = 0;
+                } else {
+                    secondQuals[secondElement.getOffset()] =
```
Declare the magic value `0.8` as a named `static final` constant at the top of the class, with a name that makes it clear the value comes from samtools, and use the constant in this method instead of the literal value.

magicDGS replied:
Done, although I guess that the name for the constant (`SAMTOOLS_OVERLAP_LOW_CONFIDENCE`) is not great. Any suggestion?

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java:
```diff
@@ -490,4 +492,124 @@ public void testSimplePileupWithOffset(){
+        Assert.assertEquals(copy, read);
+        Assert.assertEquals(copy2, read2);
+    }
+
```
Add at least one unit test that calls `ReadPileup.fixOverlaps()`.

magicDGS replied:
I use exactly the same test `DataProvider` to create a `ReadPileup` with the two `PileupElement` and call `ReadPileup.fixOverlaps()`.  Is that enought?

droazen replied:
Yes, that should be fine

magicDGS commented:
Rebase/squash and addressed all comments. `CLOUD=mandatory` is not going to pass, but the no-cloud test will. Back to you @droazen. Thanks a lot for reviewing @droazen and @fleharty.

droazen commented:
:+1: merging
